## Ejercicios

Importar matplotlib y pandas:

In [2]:
import pandas as pd

## Ejercicio 1:

Utilizar pandas para importar los datos del archivo "certificado.csv" en la variable "certificados"

In [2]:

Certificado = pd.read_csv("./certificado.csv")
Certificado.head()

nombre   examen_diagnostico  examen_final grupo maestro
0  Alejandra                 7.7           9.0     A   Maria
1    Alberto                 7.3           8.8     B   Pedro
2      César                 7.5           9.1     B   Pedro
3    Gustavo                 7.1           8.6     B   Pedro
4    Ernesto                 7.8           9.2     A   Maria

Usar groupby para crear un DataFrame donde se tenga el promedio del examen diagnostico y examen final por cada grupo, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act1Img1.png)

In [3]:
Certificado.groupby("grupo").agg({"examen_diagnostico":"mean","examen_final":"mean"})

examen_diagnostico  examen_final
grupo                                  
A                7.420000      8.800000
B                7.123077      8.592308
C                7.558824      9.023529

![Resuelto](./Img_24_Ejercicios/Act1Img2.png)

![Resuelto](./Img_24_Ejercicios/Act1Img3.png)

__Conclusión:__ todos los grupos presentan una mejora en la calificación al finalizar el curso

## Ejercicio 2:
Crear una columna llamada Progreso en el DataFrame certificados que contenga la diferencia entre el examen final y el examen diagnóstico, por ejemplo, si un alumno tuvo un 8 en el examen final y un 7 en el examen diagnóstico, la columna Progreso tendrá el valor de  1, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act2Img1.png)

In [6]:
Certificado["Progreso"] = round(Certificado["examen_final"] - Certificado["examen_diagnostico"],2)

In [7]:
Certificado.head()

nombre   examen_diagnostico  examen_final grupo maestro  Progreso
0  Alejandra                 7.7           9.0     A   Maria       1.3
1    Alberto                 7.3           8.8     B   Pedro       1.5
2      César                 7.5           9.1     B   Pedro       1.6
3    Gustavo                 7.1           8.6     B   Pedro       1.5
4    Ernesto                 7.8           9.2     A   Maria       1.4

![Resuelto](./Img_24_Ejercicios/Act2Img2.png)

__Conclusión__: la mayoria de los estudiantes tienen una mejora entre 1 y 2.5 puntos

## Ejercicio 3: 
Utilizar gruopby para obtener el progreso promedio en las calificaciones de los alumnos por cada profesor, por ejemplo, los alumnos del profesor Pedro al inicio tienen un promedio de 7, al terminar el curso aumenta a 9 y su progreso es de 2 puntos, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act3Img1.png)

In [9]:
Certificado.groupby("maestro").agg({"examen_diagnostico":"mean","examen_final":"mean","Progreso":"mean"}).round(6)

examen_diagnostico  examen_final  Progreso
maestro                                             
Gabriela            7.558824      9.023529  1.464706
Maria               7.420000      8.800000  1.380000
Pedro               7.123077      8.592308  1.469231

![Resuelto](./Img_24_Ejercicios/Act3Img2.png)

__Conclusión:__ Los 3 maestros tienen un desempeño similar.

## Ejercicio 4:
Utilizar pandas para importar los datos del archivo "certificado_historico.csv" en la variable "historico"

In [1]:
import pandas as pd 

historico = pd.read_csv("./certificado_historico.csv")
historico

num_control    nombre  edad  calificacion grupo  anio  semestre  \
0          A001     Aarón    14           8.9     A  2017         1   
1          A002     Lucía    42           9.3     A  2017         1   
2          A003      Hugo    29           8.7     A  2017         1   
3          A004   Julieta    34           7.9     A  2017         1   
4          A005    Adrían    14           9.1     A  2017         1   
..          ...       ...   ...           ...   ...   ...       ...   
295        A296    Brayan    25           9.1     C  2020         2   
296        A297    Deymar    30           9.1     C  2020         2   
297        A298   Rogelio    15           7.9     C  2020         2   
298        A299     Percy    32           7.4     C  2020         2   
299        A300  Claridad    25           7.2     C  2020         2   

    certificacion  maestro  
0              NO  Tatiana  
1              NO  Tatiana  
2              NO  Tatiana  
3              NO  Tatiana  
4              NO  Tatiana  
..            ...      ...  
295            SI   Gaspar  
296            SI   Gaspar  
297            NO   Gaspar  
298            NO   Gaspar  
299            SI   Gaspar  

[300 rows x 9 columns]

Utlizar groupby para crear un DataFrame con la calificación más baja en cada grupo de cada año, el resultado debe ser el siguiente: 

![Resuelto](./Img_24_Ejercicios/Act4Img1.png)

In [4]:
historico.groupby(["anio","grupo"]).agg({"calificacion":"min"})

calificacion
anio grupo              
2017 A               7.0
     B               7.1
     C               7.1
2018 A               7.4
     B               7.2
     C               7.0
2019 A               6.6
     B               7.1
     C               7.0
2020 A               7.1
     B               7.0
     C               7.2

![Resuelto](./Img_24_Ejercicios/Act4Img2.png)

__Conclusión:__ la calificación más baja normalmente se mantiene por encima de 7

## Ejercicio 5:
Utilizar groupby para crear un DataFrame con la cantidad de alumnos a la que ha impartido clases cada maestro, por ejemplo, la profesora Jessica ha dado clases a 80 alumnos, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act5Img1.png)

In [14]:
historico.groupby("maestro").agg({"nombre":"size"})

nombre
maestro        
Armando     100
Emily        74
Gaspar       26
Tatiana     100

![Resuelto](./Img_24_Ejercicios/Act5Img2.png)

__Conclusión:__ Tatiana y Armando son los profesores que más alumnos han tenido

## Ejercicio 6:
Utilizar pivot table para crear un DataFrame con la calificación promedio de los estudiantes de cada profesor en cada año, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act6Img1.png)

In [15]:
historico.pivot_table(index="anio",values="calificacion",columns="maestro",aggfunc="mean")

maestro  Armando     Emily    Gaspar  Tatiana
anio                                         
2017       8.552  8.368000       NaN    8.416
2018       8.380  8.604000       NaN    8.780
2019       8.728  8.608333  8.369231    8.296
2020       8.496  9.025000  8.184615    8.476

![Resuelto](./Img_24_Ejercicios/Act6Img2.png)

__Conclusión:__ Emily es la maestra que mayor progreso ha tenido

## Ejercicio 7:
Utilizar groupby para crear un DataFrame con la mediana en la calificación de cada grupo en cada año, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act7Img1.png)

In [16]:
historico.groupby(["grupo","anio"]).agg({"calificacion":"median"})

calificacion
grupo anio              
A     2017           8.1
      2018           8.7
      2019           8.2
      2020           8.3
B     2017           8.8
      2018           8.2
      2019           8.7
      2020           8.7
C     2017           8.3
      2018           8.6
      2019           8.8
      2020           8.7

__Conclusión:__ la mediana en las calificaciones normalmente es alrededor de 8.5

## Ejercicio 8
Armar un código que devuelva un Dataframe con los datos de los estudiantes que si quieren obtener un certificado, el resultado debe ser el siguiente: 

![Resuelto](./Img_24_Ejercicios/Act8Img1.png)

In [2]:
alum_certificados= historico.loc[historico["certificacion"] == "SI"]

In [6]:
alum_certificados

num_control      nombre  edad  calificacion grupo  anio  semestre  \
5          A006      Jacobo    44           9.8     A  2017         1   
7          A008       Sofía    20           9.0     A  2017         1   
9          A010     Martina    35           9.5     A  2017         1   
11         A012     Ricardo    34           9.0     A  2017         1   
12         A013      Javier    31           9.7     A  2017         2   
..          ...         ...   ...           ...   ...   ...       ...   
292        A293        Edir    23           7.7     C  2020         2   
294        A295  Cuauhtemoc    14           7.9     C  2020         2   
295        A296      Brayan    25           9.1     C  2020         2   
296        A297      Deymar    30           9.1     C  2020         2   
299        A300    Claridad    25           7.2     C  2020         2   

    certificacion  maestro  
5              SI  Tatiana  
7              SI  Tatiana  
9              SI  Tatiana  
11             SI  Tatiana  
12             SI  Tatiana  
..            ...      ...  
292            SI   Gaspar  
294            SI   Gaspar  
295            SI   Gaspar  
296            SI   Gaspar  
299            SI   Gaspar  

[115 rows x 9 columns]

Armar un código que devuelva un Dataframe con los datos de los estudiantes que no quieren obtener un certificado,el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act8Img2.png)

In [3]:
alum_NO_certificados = historico.loc[historico["certificacion"] == "NO"]
alum_NO_certificados

num_control   nombre  edad  calificacion grupo  anio  semestre  \
0          A001    Aarón    14           8.9     A  2017         1   
1          A002    Lucía    42           9.3     A  2017         1   
2          A003     Hugo    29           8.7     A  2017         1   
3          A004  Julieta    34           7.9     A  2017         1   
4          A005   Adrían    14           9.1     A  2017         1   
..          ...      ...   ...           ...   ...   ...       ...   
288        A289    Eddie    34           7.9     C  2020         2   
291        A292  Delfino    23           9.1     C  2020         2   
293        A294  Arnoldo    20           9.0     C  2020         2   
297        A298  Rogelio    15           7.9     C  2020         2   
298        A299    Percy    32           7.4     C  2020         2   

    certificacion  maestro  
0              NO  Tatiana  
1              NO  Tatiana  
2              NO  Tatiana  
3              NO  Tatiana  
4              NO  Tatiana  
..            ...      ...  
288            NO   Gaspar  
291            NO   Gaspar  
293            NO   Gaspar  
297            NO   Gaspar  
298            NO   Gaspar  

[185 rows x 9 columns]

![Resuelto](./Img_24_Ejercicios/Act8Img4.png)

__Conclusión:__ La mayoria de los estudiantes que quieren obtener un certificado tienen altas calificaciones

## Ejercicio 9:
Utilizar pivot table crear un DataFrame que contenga la calificación promedio de los estudiantes en cada año, separado por si quieren obtener un certificado o no, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act9Img1.png)

In [8]:
historico.pivot_table(index="anio",values="calificacion",columns="certificacion",aggfunc="mean").round(6)

certificacion        NO        SI
anio                             
2017           8.342857  8.747368
2018           8.315556  8.996667
2019           8.226190  8.854545
2020           8.300000  8.800000

Armar un código que devuelva un DataFrame con la diferencia en cada año respecto a la calificación promedio de los estudiantes que si querian obtener un certificado contra los que no, por ejemplo, si en el 2017 el promedio de los estudiantes que si querían un certificado era 9 y de los que no era 8, la diferencia será 1 para esa fila, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act9Img2.png)

In [4]:
diferencia_certificacion = historico.pivot_table(index="anio",columns="certificacion",values="calificacion",aggfunc="mean",fill_value=0)

In [5]:
diferencia_certificacion["DIFERENCIA"] = diferencia_certificacion["SI"] - diferencia_certificacion["NO"]
diferencia_certificacion

certificacion        NO        SI  DIFERENCIA
anio                                         
2017           8.342857  8.747368    0.404511
2018           8.315556  8.996667    0.681111
2019           8.226190  8.854545    0.628355
2020           8.300000  8.800000    0.500000

![Resuelto](./Img_24_Ejercicios/Act9Img3.png)

__Conclusión:__ los estudiantes que buscan obtener su certificado tienen mayores calificaciones en promedio

## Ejercicio 10:
Armar un código que devuelva un dataframe con las diferencias de las calificaciones de cada alumno con respecto al promedio general, por ejemplo, si Juan sacó 8 y el promedio general es 9, la diferencia será -1, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act10Img1.png)

In [6]:
promedio_general = historico.calificacion.mean()

In [7]:
def diferencia_x(x):
    return x - promedio_general

historico["diferencia"] = historico.calificacion.apply(diferencia_x).round(3)

In [8]:
historico.head(5)

num_control   nombre  edad  calificacion grupo  anio  semestre  \
0        A001    Aarón    14           8.9     A  2017         1   
1        A002    Lucía    42           9.3     A  2017         1   
2        A003     Hugo    29           8.7     A  2017         1   
3        A004  Julieta    34           7.9     A  2017         1   
4        A005   Adrían    14           9.1     A  2017         1   

  certificacion  maestro  diferencia  
0            NO  Tatiana       0.386  
1            NO  Tatiana       0.786  
2            NO  Tatiana       0.186  
3            NO  Tatiana      -0.614  
4            NO  Tatiana       0.586

Nota: se puede hacer como en el ejercicio anterior, pero se hizo diferente para mostrar más opciones a los alumnos

Armar un código que devuelva un Dataframe con los datos de los estudiantes que si quieren obtener un certificado,el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act10Img2.png)

In [9]:
historico.loc[historico["certificacion"] == "SI"]

num_control      nombre  edad  calificacion grupo  anio  semestre  \
5          A006      Jacobo    44           9.8     A  2017         1   
7          A008       Sofía    20           9.0     A  2017         1   
9          A010     Martina    35           9.5     A  2017         1   
11         A012     Ricardo    34           9.0     A  2017         1   
12         A013      Javier    31           9.7     A  2017         2   
..          ...         ...   ...           ...   ...   ...       ...   
292        A293        Edir    23           7.7     C  2020         2   
294        A295  Cuauhtemoc    14           7.9     C  2020         2   
295        A296      Brayan    25           9.1     C  2020         2   
296        A297      Deymar    30           9.1     C  2020         2   
299        A300    Claridad    25           7.2     C  2020         2   

    certificacion  maestro  diferencia  
5              SI  Tatiana       1.286  
7              SI  Tatiana       0.486  
9              SI  Tatiana       0.986  
11             SI  Tatiana       0.486  
12             SI  Tatiana       1.186  
..            ...      ...         ...  
292            SI   Gaspar      -0.814  
294            SI   Gaspar      -0.614  
295            SI   Gaspar       0.586  
296            SI   Gaspar       0.586  
299            SI   Gaspar      -1.314  

[115 rows x 10 columns]

Armar un código que devuelva un Dataframe con los datos de los estudiantes que no quieren obtener un certificado,el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act10Img3.png)

In [10]:
historico.loc[historico["certificacion"] == "NO"]

num_control   nombre  edad  calificacion grupo  anio  semestre  \
0          A001    Aarón    14           8.9     A  2017         1   
1          A002    Lucía    42           9.3     A  2017         1   
2          A003     Hugo    29           8.7     A  2017         1   
3          A004  Julieta    34           7.9     A  2017         1   
4          A005   Adrían    14           9.1     A  2017         1   
..          ...      ...   ...           ...   ...   ...       ...   
288        A289    Eddie    34           7.9     C  2020         2   
291        A292  Delfino    23           9.1     C  2020         2   
293        A294  Arnoldo    20           9.0     C  2020         2   
297        A298  Rogelio    15           7.9     C  2020         2   
298        A299    Percy    32           7.4     C  2020         2   

    certificacion  maestro  diferencia  
0              NO  Tatiana       0.386  
1              NO  Tatiana       0.786  
2              NO  Tatiana       0.186  
3              NO  Tatiana      -0.614  
4              NO  Tatiana       0.586  
..            ...      ...         ...  
288            NO   Gaspar      -0.614  
291            NO   Gaspar       0.586  
293            NO   Gaspar       0.486  
297            NO   Gaspar      -0.614  
298            NO   Gaspar      -1.114  

[185 rows x 10 columns]

![Resuelto](./Img_24_Ejercicios/Act10Img6.png)

__Conclusión:__ Los estudiantes que desean obtener un certificado tienden a tener calificaciones mayores al promedio

## Ejercicio 11
Armar un Dataframe con los indices de los alumnos que obtuvieron la calificación más alta en cada grupo en cada año, el resultado debe ser:

![Resuelto](./Img_24_Ejercicios/Act11Img1.png)

In [23]:

index_alto = historico.groupby(["anio","grupo"]).agg({"calificacion":"idxmax"})

In [24]:
index_alto

calificacion
anio grupo              
2017 A                 5
     B                26
     C                53
2018 A                86
     B               123
     C               126
2019 A               170
     B               177
     C               207
2020 A               234
     B               265
     C               278

Utilizar los indices del paso anterior para obtener en un DataFrame los datos de los estudiantes con mejor promedio empleando loc, el resultado debe ser:

![Resuelto](./Img_24_Ejercicios/Act11Img2.png)

In [28]:
mejores_promedios = historico.loc[index_alto["calificacion"]]

mejores_promedios.head()

num_control    nombre  edad  calificacion grupo  anio  semestre  \
5          A006    Jacobo    44           9.8     A  2017         1   
26         A027   Roberto    28           9.8     B  2017         1   
53         A054    Daniel    15           9.8     C  2017         1   
86         A087     Kevin    16          10.0     A  2018         1   
123        A124  Isabella    22           9.9     B  2018         2   

    certificacion  maestro  diferencia  
5              SI  Tatiana       1.286  
26             NO  Armando       1.286  
53             NO    Emily       1.286  
86             SI  Tatiana       1.486  
123            NO  Armando       1.386

![Resuelto](./Img_24_Ejercicios/Act11Img3.png)

__Conclusión:__ en todos los grupos siempre un estudiante tuvo al menos 9.8 de calificación

## Ejercicio 12
Agregar una columna llamada id al DataFrame anterior , formada por la unión de la columna anio y la columna grupo, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act12Img1.png)

In [30]:
mejores_promedios["id"] = mejores_promedios["anio"].astype(str)+mejores_promedios["grupo"]

In [31]:
mejores_promedios

num_control    nombre  edad  calificacion grupo  anio  semestre  \
5          A006    Jacobo    44           9.8     A  2017         1   
26         A027   Roberto    28           9.8     B  2017         1   
53         A054    Daniel    15           9.8     C  2017         1   
86         A087     Kevin    16          10.0     A  2018         1   
123        A124  Isabella    22           9.9     B  2018         2   
126        A127     Paola    12          10.0     C  2018         1   
170        A171     Edson    16           9.9     A  2019         2   
177        A178    Matias    31          10.0     B  2019         1   
207        A208      Jaan    30           9.8     C  2019         1   
234        A235    Rodney    38          10.0     A  2020         1   
265        A266    Marwin    16          10.0     B  2020         2   
278        A279  Penelope    23           9.9     C  2020         1   

    certificacion  maestro  diferencia     id  
5              SI  Tatiana       1.286  2017A  
26             NO  Armando       1.286  2017B  
53             NO    Emily       1.286  2017C  
86             SI  Tatiana       1.486  2018A  
123            NO  Armando       1.386  2018B  
126            SI    Emily       1.486  2018C  
170            SI  Tatiana       1.386  2019A  
177            SI  Armando       1.486  2019B  
207            NO    Emily       1.286  2019C  
234            SI  Tatiana       1.486  2020A  
265            NO  Armando       1.486  2020B  
278            SI    Emily       1.386  2020C

![Resuelto](./Img_24_Ejercicios/Act12Img2.png)

__Conclusión:__ en todos los grupos siempre un estudiante tuvo al menos 9.8 de calificación

## Ejercicio 13
Utilizar pandas para importar los datos del archivo "hospital.xlsx" en la variable "hospital"

In [3]:
import pandas as pd
import datetime as dt

In [2]:
hospital = pd.read_excel("./hospital.xlsx")
hospital

Fecha                     Nombre                    Doctor  \
0    2018-01-01    Bastian Hurtado Cabrero      Simón Castro Infante   
1    2018-01-02         Carla Rosas Ávalos      Ismael López Saldaña   
2    2018-01-03   Natanael Lazcano Aguilar       Félix Llanos Macías   
3    2018-01-04     Judith Hinojosa Torres      Ismael López Saldaña   
4    2018-01-07     Giovanni Rudomín Arias  Paulina Contreras Cuevas   
...         ...                        ...                       ...   
3595 2021-12-16   Jaqueline Burgos Meneses      Nestor Núñez Basurto   
3596 2021-12-25     Andrea  Larrea Borundo      Simón Castro Infante   
3597 2021-12-26  Valentina Morales Márquez         Alfonso Soda Neri   
3598 2021-12-26   Casandra Saturno Vázquez    Arnoldo Ondarza Hernán   
3599 2021-12-28    Gleidin Urbina Iglesias      Nestor Núñez Basurto   

             Especialidad       Tipo  Honorarios  Calificación  Comisión  
0             Hematología   Consulta        1025          0.84      0.16  
1             Hematología   Consulta        1025          0.83      0.11  
2            Cardiología    Consulta        1075          0.78      0.12  
3             Hematología   Consulta        1075          0.92      0.17  
4      Gastroenterología.   Consulta        1125          0.89      0.12  
...                   ...        ...         ...           ...       ...  
3595          Hematología  Operación       39500          0.89      0.16  
3596  Cirugía Pediátrica.  Operación       18000          0.80      0.06  
3597     Cirugía Plástica  Operación       38500          0.62      0.20  
3598          Ginegología  Operación       46500          0.83      0.19  
3599          Hematología  Operación       41000          0.75      0.05  

[3600 rows x 8 columns]

Agregar una columna llamada "FechaMes", que contenga el año y el mes de la columna fecha (Ej. 2010-03)

![Resuelto](./Img_24_Ejercicios/Act13Img1.png)

In [5]:
hospital["FechaMes"] = hospital.Fecha.dt.strftime("%Y-%m")
hospital

Fecha                     Nombre                    Doctor  \
0    2018-01-01    Bastian Hurtado Cabrero      Simón Castro Infante   
1    2018-01-02         Carla Rosas Ávalos      Ismael López Saldaña   
2    2018-01-03   Natanael Lazcano Aguilar       Félix Llanos Macías   
3    2018-01-04     Judith Hinojosa Torres      Ismael López Saldaña   
4    2018-01-07     Giovanni Rudomín Arias  Paulina Contreras Cuevas   
...         ...                        ...                       ...   
3595 2021-12-16   Jaqueline Burgos Meneses      Nestor Núñez Basurto   
3596 2021-12-25     Andrea  Larrea Borundo      Simón Castro Infante   
3597 2021-12-26  Valentina Morales Márquez         Alfonso Soda Neri   
3598 2021-12-26   Casandra Saturno Vázquez    Arnoldo Ondarza Hernán   
3599 2021-12-28    Gleidin Urbina Iglesias      Nestor Núñez Basurto   

             Especialidad       Tipo  Honorarios  Calificación  Comisión  \
0             Hematología   Consulta        1025          0.84      0.16   
1             Hematología   Consulta        1025          0.83      0.11   
2            Cardiología    Consulta        1075          0.78      0.12   
3             Hematología   Consulta        1075          0.92      0.17   
4      Gastroenterología.   Consulta        1125          0.89      0.12   
...                   ...        ...         ...           ...       ...   
3595          Hematología  Operación       39500          0.89      0.16   
3596  Cirugía Pediátrica.  Operación       18000          0.80      0.06   
3597     Cirugía Plástica  Operación       38500          0.62      0.20   
3598          Ginegología  Operación       46500          0.83      0.19   
3599          Hematología  Operación       41000          0.75      0.05   

     FechaMes  
0     2018-01  
1     2018-01  
2     2018-01  
3     2018-01  
4     2018-01  
...       ...  
3595  2021-12  
3596  2021-12  
3597  2021-12  
3598  2021-12  
3599  2021-12  

[3600 rows x 9 columns]

Agregar una columna llamada "Año", que contenga el año y el mes de la columna fecha (Ej. 2010)

![Resuelto](./Img_24_Ejercicios/Act13Img2.png)

In [9]:
hospital["Año"] = hospital.Fecha.dt.year

hospital

Fecha                     Nombre                    Doctor  \
0    2018-01-01    Bastian Hurtado Cabrero      Simón Castro Infante   
1    2018-01-02         Carla Rosas Ávalos      Ismael López Saldaña   
2    2018-01-03   Natanael Lazcano Aguilar       Félix Llanos Macías   
3    2018-01-04     Judith Hinojosa Torres      Ismael López Saldaña   
4    2018-01-07     Giovanni Rudomín Arias  Paulina Contreras Cuevas   
...         ...                        ...                       ...   
3595 2021-12-16   Jaqueline Burgos Meneses      Nestor Núñez Basurto   
3596 2021-12-25     Andrea  Larrea Borundo      Simón Castro Infante   
3597 2021-12-26  Valentina Morales Márquez         Alfonso Soda Neri   
3598 2021-12-26   Casandra Saturno Vázquez    Arnoldo Ondarza Hernán   
3599 2021-12-28    Gleidin Urbina Iglesias      Nestor Núñez Basurto   

             Especialidad       Tipo  Honorarios  Calificación  Comisión  \
0             Hematología   Consulta        1025          0.84      0.16   
1             Hematología   Consulta        1025          0.83      0.11   
2            Cardiología    Consulta        1075          0.78      0.12   
3             Hematología   Consulta        1075          0.92      0.17   
4      Gastroenterología.   Consulta        1125          0.89      0.12   
...                   ...        ...         ...           ...       ...   
3595          Hematología  Operación       39500          0.89      0.16   
3596  Cirugía Pediátrica.  Operación       18000          0.80      0.06   
3597     Cirugía Plástica  Operación       38500          0.62      0.20   
3598          Ginegología  Operación       46500          0.83      0.19   
3599          Hematología  Operación       41000          0.75      0.05   

     FechaMes   Año  
0     2018-01  2018  
1     2018-01  2018  
2     2018-01  2018  
3     2018-01  2018  
4     2018-01  2018  
...       ...   ...  
3595  2021-12  2021  
3596  2021-12  2021  
3597  2021-12  2021  
3598  2021-12  2021  
3599  2021-12  2021  

[3600 rows x 10 columns]

Crear una función utilizando groupby que devuelva un DataFrame con el total de visitas que si hicieron al hospital y el promedio de los honorarios en cada año, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act13Img4.png)

In [13]:
hospital.groupby("Año").agg({"Honorarios":["mean","count"]})

Honorarios      
             mean count
Año                    
2018  5689.717295   902
2019  6040.718232   905
2020  5869.656284   931
2021  6139.675174   862

![Resuelto](./Img_24_Ejercicios/Act13Img5.png)

![Resuelto](./Img_24_Ejercicios/Act13Img6.png)

__Conclusión:__ el costo promedio de los honorarios y el número de visitas no han tenido grandes variaciones.

## Ejercicio 14
Crear una función que devuelva un DataFrame con los registros de las visitas al hospital que fueron con motivo de una consulta médica, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act14Img1.png)

In [17]:
consulta_medica = hospital.loc[hospital["Tipo"] == "Consulta"]

In [18]:
consulta_medica

Fecha                       Nombre                    Doctor  \
0    2018-01-01      Bastian Hurtado Cabrero      Simón Castro Infante   
1    2018-01-02           Carla Rosas Ávalos      Ismael López Saldaña   
2    2018-01-03     Natanael Lazcano Aguilar       Félix Llanos Macías   
3    2018-01-04       Judith Hinojosa Torres      Ismael López Saldaña   
4    2018-01-07       Giovanni Rudomín Arias  Paulina Contreras Cuevas   
...         ...                          ...                       ...   
1948 2021-12-23           Miguel Mejía Arias   Constanza Cetina Cuenca   
1949 2021-12-23       Yaneth Revuelta Ávalos     Gonzalo Ondarza Gómez   
1950 2021-12-25          Nelson Molina Funes   Óscar Richardson Ávalos   
1951 2021-12-26       Javier Cervera Lazcano      Ismael López Saldaña   
1952 2021-12-27  William Contreras Santillas     Gonzalo Ondarza Gómez   

            Especialidad      Tipo  Honorarios  Calificación  Comisión  \
0            Hematología  Consulta        1025          0.84      0.16   
1            Hematología  Consulta        1025          0.83      0.11   
2           Cardiología   Consulta        1075          0.78      0.12   
3            Hematología  Consulta        1075          0.92      0.17   
4     Gastroenterología.  Consulta        1125          0.89      0.12   
...                  ...       ...         ...           ...       ...   
1948         Hematología  Consulta        1000          0.71      0.10   
1949         Ginegología  Consulta        1150          0.71      0.17   
1950    Cirugía General.  Consulta        1050          0.86      0.05   
1951         Hematología  Consulta        1100          0.80      0.12   
1952         Ginegología  Consulta        1000          0.89      0.07   

     FechaMes   Año  
0     2018-01  2018  
1     2018-01  2018  
2     2018-01  2018  
3     2018-01  2018  
4     2018-01  2018  
...       ...   ...  
1948  2021-12  2021  
1949  2021-12  2021  
1950  2021-12  2021  
1951  2021-12  2021  
1952  2021-12  2021  

[1953 rows x 10 columns]

![Resuelto](./Img_24_Ejercicios/Act14Img2.png)

__Conclusión:__ el costo por una consulta normalmente es de 1000 o 1250 pesos

## Ejercicio 15
Crear una función utilizando groupby que devuelva un DataFrame con el total de visitas al hospital por cada especialidad, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act15Img1.png)

In [22]:
hospital.groupby("Especialidad").agg({"Doctor":"size"})

Doctor
Especialidad               
Cardiología             390
Cirugía General.        438
Cirugía Pediátrica.     294
Cirugía Plástica        549
Dermatología            651
Gastroenterología.      285
Ginegología             312
Hematología             522
Nefrología              159

![Resuelto](./Img_24_Ejercicios/Act15Img2.png)

__Conclusión:__ la especialidad más demandada es la dermatología

## Ejercicio 16

Agregar una columna llamada "Mes", que contenga el año y el mes de la columna fecha (Ej. 01)

![Resuelto](./Img_24_Ejercicios/Act16Img1.png)

In [46]:
hospital["Mes"] = hospital.Fecha.dt.month

hospital.head()

Fecha                    Nombre                    Doctor  \
0 2018-01-01   Bastian Hurtado Cabrero      Simón Castro Infante   
1 2018-01-02        Carla Rosas Ávalos      Ismael López Saldaña   
2 2018-01-03  Natanael Lazcano Aguilar       Félix Llanos Macías   
3 2018-01-04    Judith Hinojosa Torres      Ismael López Saldaña   
4 2018-01-07    Giovanni Rudomín Arias  Paulina Contreras Cuevas   

         Especialidad      Tipo  Honorarios  Calificación  Comisión FechaMes  \
0         Hematología  Consulta        1025          0.84      0.16  2018-01   
1         Hematología  Consulta        1025          0.83      0.11  2018-01   
2        Cardiología   Consulta        1075          0.78      0.12  2018-01   
3         Hematología  Consulta        1075          0.92      0.17  2018-01   
4  Gastroenterología.  Consulta        1125          0.89      0.12  2018-01   

    Año  Mes  
0  2018    1  
1  2018    1  
2  2018    1  
3  2018    1  
4  2018    1

Crear una función que devuelva un DataFrame con la información de las visitas al hospital cuyo motivo fue una operación, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act16Img2.png)

In [47]:
operacion_df = hospital.loc[hospital["Tipo"] == "Operación"]

In [48]:
operacion_df

Fecha                     Nombre                    Doctor  \
3063 2018-01-03      Stephanie Rioja Núñez  Paulina Contreras Cuevas   
3064 2018-01-05        Adán Velasco Pizano      Ismael López Saldaña   
3065 2018-01-06     Jesús Posadas Peñalosa       Félix Llanos Macías   
3066 2018-01-19       Álvaro Arroyo Varela  Paulina Contreras Cuevas   
3067 2018-01-20       Lucas Roldán Ondarza      Ismael López Saldaña   
...         ...                        ...                       ...   
3595 2021-12-16   Jaqueline Burgos Meneses      Nestor Núñez Basurto   
3596 2021-12-25     Andrea  Larrea Borundo      Simón Castro Infante   
3597 2021-12-26  Valentina Morales Márquez         Alfonso Soda Neri   
3598 2021-12-26   Casandra Saturno Vázquez    Arnoldo Ondarza Hernán   
3599 2021-12-28    Gleidin Urbina Iglesias      Nestor Núñez Basurto   

             Especialidad       Tipo  Honorarios  Calificación  Comisión  \
3063   Gastroenterología.  Operación       15000          0.88      0.07   
3064          Hematología  Operación       25000          0.96      0.06   
3065         Cardiología   Operación       22500          0.82      0.06   
3066   Gastroenterología.  Operación       37000          0.99      0.17   
3067          Hematología  Operación       12000          0.73      0.10   
...                   ...        ...         ...           ...       ...   
3595          Hematología  Operación       39500          0.89      0.16   
3596  Cirugía Pediátrica.  Operación       18000          0.80      0.06   
3597     Cirugía Plástica  Operación       38500          0.62      0.20   
3598          Ginegología  Operación       46500          0.83      0.19   
3599          Hematología  Operación       41000          0.75      0.05   

     FechaMes   Año  Mes  
3063  2018-01  2018    1  
3064  2018-01  2018    1  
3065  2018-01  2018    1  
3066  2018-01  2018    1  
3067  2018-01  2018    1  
...       ...   ...  ...  
3595  2021-12  2021   12  
3596  2021-12  2021   12  
3597  2021-12  2021   12  
3598  2021-12  2021   12  
3599  2021-12  2021   12  

[537 rows x 11 columns]

Utilizar pivot table para crear una función que devuelva un DataFrame con los honorarios de la operación más barata en cada mes de cada año, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act16Img2.png)

In [51]:
operacion_df.pivot_table(index="Fecha",values="Honorarios",aggfunc="min")

Honorarios
Fecha                 
2018-01-03       15000
2018-01-05       25000
2018-01-06       22500
2018-01-19       37000
2018-01-20       12000
...                ...
2021-12-12       30000
2021-12-16       32000
2021-12-25       18000
2021-12-26       38500
2021-12-28       41000

[436 rows x 1 columns]

![Resuelto](./Img_24_Ejercicios/Act16Img3.png)

__Conclusión:__ en el año 2021 es donde las operaciones más económicas tienen el costo más alto

## Ejercicio 17
Crear una función que utilizando groupby devuelva un DataFrame con la cantidad de visitas atendidas por cada doctor, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act17Img1.png)

In [53]:
hospital.groupby("Doctor").agg({"Doctor":"count"})

Doctor
Doctor                           
Alastor Mansilla Rivas        144
Alfonso Soda Neri             303
Arnoldo Ondarza Hernán        120
Constanza Cetina Cuenca       135
Daniel Mejía Fraga            150
Flavio Ávalos Arroyo          126
Félix Llanos Macías           180
Gonzalo Ondarza Gómez         192
Isabella Roldán Velásquez     336
Ismael López Saldaña          224
Iván Ramos Robles             159
Javier Rosado Obrador          90
Lucía Flores Muñoz            189
Mauro Funes Meneses           108
Nestor Núñez Basurto          162
Nicolás Vargas Tapia          102
Paulina Contreras Cuevas      195
Simón Castro Infante          145
Óscar Montalvo Cuevas         246
Óscar Richardson Ávalos       294

![Resuelto](./Img_24_Ejercicios/Act17Img2.png)

## Ejercicio 18
Agregar una columna al DataFrame hospital llamada ingreso, este se obtiene al multiplicar los honorarios de cada visita con el porcentaje de comisión, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act18Img1.png)

In [54]:
hospital["Ingreso"] = hospital["Honorarios"]*hospital["Comisión"]

In [55]:
hospital.head()

Fecha                    Nombre                    Doctor  \
0 2018-01-01   Bastian Hurtado Cabrero      Simón Castro Infante   
1 2018-01-02        Carla Rosas Ávalos      Ismael López Saldaña   
2 2018-01-03  Natanael Lazcano Aguilar       Félix Llanos Macías   
3 2018-01-04    Judith Hinojosa Torres      Ismael López Saldaña   
4 2018-01-07    Giovanni Rudomín Arias  Paulina Contreras Cuevas   

         Especialidad      Tipo  Honorarios  Calificación  Comisión FechaMes  \
0         Hematología  Consulta        1025          0.84      0.16  2018-01   
1         Hematología  Consulta        1025          0.83      0.11  2018-01   
2        Cardiología   Consulta        1075          0.78      0.12  2018-01   
3         Hematología  Consulta        1075          0.92      0.17  2018-01   
4  Gastroenterología.  Consulta        1125          0.89      0.12  2018-01   

    Año  Mes  Ingreso  
0  2018    1   164.00  
1  2018    1   112.75  
2  2018    1   129.00  
3  2018    1   182.75  
4  2018    1   135.00

Crear una función utilizando pivot table que devuelva un DataFrame con la suma de los ingresos en cada año por tipo de visita:

![Resuelto](./Img_24_Ejercicios/Act18Img2.png)

In [57]:
hospital.pivot_table(index="Año",values="Ingreso",aggfunc="sum",columns="Tipo")

Tipo  Consulta    Examen  Operación
Año                                
2018   69302.0  121121.0   468165.0
2019   68941.5  109276.0   498010.0
2020   70183.5  122730.0   489795.0
2021   65630.0  100555.0   466955.0

![Resuelto](./Img_24_Ejercicios/Act18Img3.png)

__Conclusión:__ El tipo de consulta que más ingresos genera al hospital son las operaciones 

## Ejercicio 19
Crear un código utilizando groupby que devuelva un DataFrame con la mediana en los ingresos en cada mes de cada año, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act19Img1.png)

![Resuelto](./Img_24_Ejercicios/Act19Img2.png)

__Conclusión:__ La mediana de los ingresos normalmente oscila entre los 150 y 200 pesos

## Ejercicio 20
Crear una función que devuelva un DataFrame con la información de las visitas al hospital cuyo motivo fue una consulta médica, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act20Img1.png)

Crear una función que devuelva un DataFrame con la información de las visitas al hospital cuyo motivo fue un examen médico:

![Resuelto](./Img_24_Ejercicios/Act20Img2.png)

![Resuelto](./Img_24_Ejercicios/Act20Img4.png)

__Conclusión:__ las consultas médicas son las que menos ingresos reportan al hospital

## Ejercicio 21
Crear una función que devuelva un DataFrame con la información de las visitas médicas que no fueron por una operación, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act21Img1.png)

Utilizando el DataFrame anterior crear una función que devuelva el ingreso promedio de todas las visitas al hospital cuyo motivo no fue una operación:

Crear una función que devuelva un DataFrame con la diferencia de los ingresos de cada visita con el promedio general, por ejemplo, si se tuvo 200 pesos de ingresos en una consulta, pero el promedio es 150, la diferencia será 50, el resultado debe ser el siguiente: 

![Resuelto](./Img_24_Ejercicios/Act21Img2.png)

Crear una función utilizando pivot table que devuelva un DataFrame con la diferencia promedio del ingreso en cada año y en cada tipo de visita, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act21Img3.png)

![Resuelto](./Img_24_Ejercicios/Act21Img4.png)

__Conclusión:__ los ingresos promedio de las consultas siempre son menores al promedio general contando examenes

## Ejercicio 22
Utilizando el DataFrame del ejercicio anterior, crear una función que devuelva los datos de las visitas al hospital cuyo motivo fue una consulta médica, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act22Img1.png)

Utilizando el DataFrame del ejercicio anterior, crear una función que devuelva los datos de las visitas al hospital cuyo motivo fue un examen médico, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act22Img2.png)

![Resuelto](./Img_24_Ejercicios/Act22Img5.png)

__Conclusión:__ la mayoría de las consultas médicas reportan un ingreso menor al promedio general  

## Ejercicio 23
Crear una función utilizando groupby que devuelva un DataFrame con los indices de las visitas médicas que más honorarios pagaron en cada año, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act23Img1.png)

Crear una función utilizando iloc que nos devuelva los datos de las visitas médicas que pagaron los honorarios más altos en cada año, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act23Img2.png)

![Resuelto](./Img_24_Ejercicios/Act23Img3.png)

__Conclusión:__ el honorario más alto en las operaciones normalmente es de 50,000 pesos

## Ejercicio 24:
Utilizando el DataFrame anterior, crear una función que agregue una columna formada por la unión del año y la especialidad del médico, el resultado debe ser el siguiente:

![Resuelto](./Img_24_Ejercicios/Act24Img1.png)

![Resuelto](./Img_24_Ejercicios/Act24Img2.png)

__Conclusión:__ la dermatología es la especialidad donde más veces se tuvo el honorario más alto